zebra-network/src/peer_set/candidate_set/tests/prop.rs

```diff
@@ -63,33 +112,60 @@ proptest! {
-            sleep(Duration::from_millis(400)).await;
+            sleep(MAX_TEST_CANDIDATES * MIN_PEER_CONNECTION_INTERVAL).await;
             // Check that the next peers are still respecting the rate limiting, without causing a
             // burst of reconnections
             check_candidates_rate_limiting(&mut candidate_set, extra_candidates).await;
         };
 
-        assert!(runtime.block_on(timeout(Duration::from_secs(10), checks)).is_ok());
+        // Allow enough time for the maximum number of candidates,
+        // plus some extra time for test machines with high CPU load.
+        // (The max delay asserts usually fail before hitting this timeout.)
+        let max_rate_limit_sleep = 3 * MAX_TEST_CANDIDATES * MIN_PEER_CONNECTION_INTERVAL;
+        let max_extra_delay = (2 * MAX_TEST_CANDIDATES + 1) * MAX_SLEEP_EXTRA_DELAY;
+        assert!(runtime.block_on(timeout(max_rate_limit_sleep + max_extra_delay, checks)).is_ok());
     }
 }
 
 /// Check if obtaining a certain number of reconnection peers is rate limited.
 ///
 /// # Panics
 ///
-/// Will panic if a reconnection peer is returned too quickly, or if no reconnection peer is
-/// returned.
-async fn check_candidates_rate_limiting<S>(candidate_set: &mut CandidateSet<S>, candidates: usize)
+/// Will panic if:
+/// - a connection peer is returned too quickly,
+/// - a connection peer is returned too slowly, or
+/// - if no reconnection peer is returned at all.
+async fn check_candidates_rate_limiting<S>(candidate_set: &mut CandidateSet<S>, candidates: u32)
 where
     S: tower::Service<Request, Response = Response, Error = BoxError>,
     S::Future: Send + 'static,
 {
-    let mut minimum_reconnect_instant = Instant::now();
+    let mut now = Instant::now();
+    let mut minimum_reconnect_instant = now;
+    // Allow extra time for test machines with high CPU load
+    let mut maximum_reconnect_instant = now + MAX_SLEEP_EXTRA_DELAY;
 
     for _ in 0..candidates {
-        assert!(candidate_set.next().await.is_some());
-        assert!(Instant::now() >= minimum_reconnect_instant);
-
-        minimum_reconnect_instant += MIN_PEER_CONNECTION_INTERVAL;
+        assert!(
+            candidate_set.next().await.is_some(),
+            "there are enough available candidates"
+        );
+
+        now = Instant::now();
+        assert!(
+            now >= minimum_reconnect_instant,
+            "all candidates should obey the minimum rate-limit: now: {:?} min: {:?}",
+            now,
+            minimum_reconnect_instant,
+        );
+        assert!(
+            now <= maximum_reconnect_instant,
+            "rate-limited candidates should not be delayed too long: now: {:?} max: {:?}. Hint: is the test machine overloaded?",
+            now,
+            maximum_reconnect_instant,
+        );
+
+        minimum_reconnect_instant = now + MIN_PEER_CONNECTION_INTERVAL;
+        maximum_reconnect_instant = now + MIN_PEER_CONNECTION_INTERVAL + MAX_SLEEP_EXTRA_DELAY;
```

Comment:
Two test changes:
- add a maximum time that's a second after the minimum time (unfortunately macOS tends to fail on shorter times)
- get the value of `now` once per loop iteration